### PR TITLE
10412 Design Debt: Message Table Updates

### PR DESCRIPTION
**Pull request by Mwindo**
The design debt [ticket](https://github.com/orgs/flexion/projects/11/views/1?pane=issue&itemId=76509485&issue=flexion%7Cef-cms%7C10412) is more or less for getting the sort icons in the sortable message table to not wrap independently. We also decided to allow the table to scroll horizontally as needed.

In the process, I decided it was worth embarking upon a refactor. That is why the PR is not tiny.

**Before (icon wraps independently):**
Screenshot 2024-10-28 at 9 42 23 AM

**After (icon wraps with text):**
Screenshot 2024-10-28 at 9 41 14 AM

**If reviewers do not like the refactor, I can just add changes to fix the icon wrapping, scroll the table, and wrap columns with text containing no spaces (as the current behavior is to allow them to expand horizontally without bound).**

---

Refactor notes:

_TL;DR: Instead of 6 separate message tables all independently built but using a lot of the same code, there is 1 message table that each of the 6 message tables instantiate with different configurations. The refactor is not perfect, but I think it is 1) better and 2) a good start to further refactoring._

While fixing "the" message table, I saw there were actually six very similar message tables (inbox, outbox, and completed for both individuals and sections). I do not like this: 1) every update to "the" message table needs to touch multiple files, 2) it is easy for the tables to get out of sync ("whoops, someone forgot to add that class in one of the files"), and 3) every table is independently responsible for making sure the header columns and row columns match (e.g., that we don't have off-by-one errors where someone forgot a `colspan`). Moreover, I saw that what fields one could sort by/display were opaque: to answer that question, one needed to go into each component separately.

My thoughts are that you should be able to see what fields are available in a given context, pass in some subset of them into some `SortableTable` component, pass in the data object (messages, docket entries, whatever), and see everything "just work." Especially with [10525](https://github.com/orgs/flexion/projects/11/views/1?pane=issue&itemId=82760879&issue=flexion%7Cef-cms%7C10525) and likely other stories adding sorting to tables, I want to get closer to this `SortableTable` ideal. (And then extend this to `FilteredSortableTable`.)

This PR is a step in that direction. It creates a `MessageTable` component that each of the six message tables uses, along with a helper class detailing what fields are available to sort by. My hope is that we can continue to move in this direction to make things less message-specific and therefore reusable across the app.

For transparency, here are (I believe) the pros and cons:

Pros: 

1. Changes to one table (e.g., styles) are maintained across the different message tables
2. `MessageTable` automatically creates the header rows and data rows from the passed in columns without the developer needing to coordinate them
3. `MessageColumns` makes it easier for a developer to see how to construct a table for messages
4. Code deduplication

Cons:

1. The obverse of 1 above: changes to one table affect all others.
2. There are a couple of clunky places in `MessageTable`--concessions to "these tables are all basically--but not _quite_--the same. I think there are ways to improve this: this is not a perfect refactor, just a good start.
3. There is slightly more rendering logic.

### CODE REVIEW

File: web-client/src/ustc-ui/Table/SortableColumn.tsx [hunk 1]
```diff
@@ -62,27 +62,34 @@ export const SortableColumn = ({
       }}
     >
       <span
-        className={classNames('margin-right-105', {
+        className={classNames('margin-right-1', {
```
Comment (Mwindo) [inline web-client/src/ustc-ui/Table/SortableColumn.tsx:65]: This looked nicer with the icon wrap in place.

File: web-client/src/ustc-ui/Table/SortableColumn.tsx [hunk 2]
```diff
@@ -62,27 +62,34 @@ export const SortableColumn = ({
...
+          )}
+          {!isLoading &&
+            getFontAwesomeIcon({
+              ascText,
+              descText,
+              direction: currentlySortedOrder,
+              isActiveColumn: isActive,
+            })}
+        </span>
+      </span>
```
Comment (Mwindo) [inline web-client/src/ustc-ui/Table/SortableColumn.tsx:92]: This represents the actual fix for the ticket.

File: web-client/src/views/DocketRecord/DocketRecord.tsx [hunk 3]
```diff
@@ -55,7 +55,7 @@ export const DocketRecord = connect(
         <DocketRecordHeader />
 
         <NonPhone>
-          <div style={{ overflowX: 'scroll', width: '100%' }}>
+          <div className="width-full overflow-x-auto">
```
Comment (Mwindo) [inline web-client/src/views/DocketRecord/DocketRecord.tsx:58]: Just updating in the places i saw it across the app for consistency.

File: web-client/src/views/Messages/MessageColumns.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
+import {
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageColumns.ts:1]: This is the "here are the available fields for a messages table" class. Eventually, filters should go here too ... but one step at a time.

My hope is that we can pull out and reuse the message-independent stuff in order to standardize our sortable tables.

File: web-client/src/views/Messages/MessageList.tsx [hunk 5]
```diff
@@ -0,0 +1,396 @@
+import {
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageList.tsx:1]: This is "the" message table. You pass in what columns and filters you want, and it takes care of the rest.

Eventually, I would love to see a message-independent "you pass in what columns, filters, and data you want (maybe with specific components for specific cells), and it takes care of the rest" class.

File: web-client/src/views/Messages/MessageList.tsx [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+};
+
+const MessageRow = ({
+  columns,
+  message,
+  messageListId,
+  onSelect,
+  selectable,
+}: {
+  message: any;
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageList.tsx]: I will try to type this better if there is an existing type that works.

File: web-client/src/views/Messages/MessagesIndividualCompleted.tsx [hunk 7]
```diff
@@ -1,156 +1,29 @@
-import { Button } from '../../ustc-ui/Button/Button';
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessagesIndividualCompleted.tsx:1]: Here, as in `MessagesIndividualInbox`, `MessagesSectionOutbox`, etc., we just pass in the columns and filters we want into `MessageList`.

File: cypress/local-only/tests/integration/messages/messages.cy.ts [hunk 8]
```diff
@@ -92,79 +92,18 @@ describe('Messages', () => {
...
-          ).click();
-          verifySubjectTitleOrder({
-            boxType: 'inbox',
-            isAscending: false,
-            prefix: 'Subject Line',
-            queueType: 'individual',
-          });
-        });
+      beforeEach(() => {
+        cy.visit('/');
```
Comment (Mwindo) [inline cypress/local-only/tests/integration/messages/messages.cy.ts:100]: Not logging in before each sort test (we don't need too--we just need to navigate to the home screen) shaved about 15s (~12%) off of this test file.

File: web-client/src/views/Messages/MessageTable.tsx [hunk 9]
```diff
@@ -0,0 +1,394 @@
+import {
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageTable.tsx:1]: Here, as in `MessagesIndividualInbox`, `MessagesSectionOutbox`, etc., we just pass in the columns and filters we want into `MessageList`.

Comment (Mwindo) [inline web-client/src/views/Messages/MessageTable.tsx:1]: This is "the" message table. You pass in what columns and filters you want, and it takes care of the rest.

Eventually, I would love to see a message-independent "you pass in what columns, filters, and data you want (maybe with specific components for specific cells), and it takes care of the rest" class.

File: web-client/src/views/Messages/MessageColumns.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  DOCKET_NUMBER: {
+    displayField: 'docketNumberWithSuffix',
+    sortField: 'docketNumber',
+    sortType: 'string',
+  },
+  FROM: { sortField: 'from', sortType: 'string' },
+  FROM_SECTION: { sortField: 'fromSectionFormatted', sortType: 'string' },
+  SUBJECT: { sortField: 'subject', sortType: 'string' },
+  TO: { sortField: 'to', sortType: 'string' },
+  TO_SECTION: { sortField: 'toSection', sortType: 'string' },
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageColumns.ts:47]: Curiously, we don't have a `toSectionFormatted` (so `X's Chambers` is rendered `xsChambers`). If this is easy to add, I'll consider it.

File: web-client/src/views/Messages/MessageTable.tsx [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+            role="alert"
+          >
+            <div className="usa-alert__body">
+              Message(s) completed at{' '}
+              {messagesIndividualInboxHelper.messagesCompletedAt} by{' '}
+              {messagesIndividualInboxHelper.messagesCompletedBy}
+            </div>
+          </div>
+        )}
+        <div className="overflow-x-auto" id={id}>
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageTable.tsx:145]: Making message tables scrollable.

File: web-client/src/views/Messages/MessageTable.tsx [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+          <div className="grid-row grid-gap">
+            {messageFilters.length > 0 && (
+              <div className="desktop:grid-col-8 tablet:grid-col-12 display-flex flex-align-center">
+                <TableFilters
+                  filters={messageFilters}
+                  onSelect={updateMessageFilterSequence}
+                ></TableFilters>
+              </div>
+            )}
+            {selectable && getCompleteAllButton()}
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageTable.tsx:155]: Coupling selectable and completable ... but for now, I think that's ok. We can separate them if needed.

File: web-client/src/views/Messages/MessageTable.tsx [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+          </div>
+
+          <table className="usa-table ustc-table subsection">
+            <thead>
+              <tr>
+                {selectable && getSelectAllCheckbox()}
+                {messageColumns.map(columnData => {
```
Comment (Mwindo) [inline web-client/src/views/Messages/MessageTable.tsx:162]: High-level: for every passed-in column, construct the headers. Then, for every data point, construct the row based on the same passed-in columns.

**Review by akuny — APPROVED** [review]
Looks like a step in the right direction!

**Pull request merged**